forecasting/src/autogluon/forecasting/dataset/ts_dataframe.py

```diff
@@ -0,0 +1,331 @@
+                        2019-01-03       2
+                1       2019-01-01       3
+                        2019-01-02       4
+                        2019-01-03       5
+                2       2019-01-01       6
+                        2019-01-02       7
+                        2019-01-03       8
+                This example can be found using example() function.
+    """
+
+    def __init__(self, data: Any, *args, **kwargs):
+        if isinstance(data, TimeSeriesListDataset):
+            data = self.from_list_dataset(data)
+        if isinstance(data, pd.DataFrame):
+            if not isinstance(data.index, pd.MultiIndex):
+                data = self.from_data_frame(data)
+        self._validate_multi_index_data_frame(data)
+        super().__init__(data=data, *args, **kwargs)
+
+    @classmethod
+    def example(cls):
+        """An example TimeSeriesDataFrame.
+
+        Returns
+        -------
+        ts_df : TimeSeriesDataFrame
+            It returns an example TimeSeriesDataFrame as:
+                                    target
+            item_id timestamp
+            0       2019-01-01       0
+                    2019-01-02       1
+                    2019-01-03       2
+            1       2019-01-01       3
+                    2019-01-02       4
+                    2019-01-03       5
+            2       2019-01-01       6
+                    2019-01-02       7
+                    2019-01-03       8
+        """
+
+        target = np.arange(9)
+        datetime_index = tuple(
+            pd.date_range(pd.Timestamp("01-01-2019"), periods=3, freq="D")
+        )
+        item_ids = (0, 1, 2)
+        multi_index = pd.MultiIndex.from_product(
+            [item_ids, datetime_index], names=["item_id", "timestamp"]
+        )
+        return TimeSeriesDataFrame(
+            pd.Series(target, name="target", index=multi_index).to_frame()
+        )
+
+    @classmethod
+    def _validate_data_frame(cls, df: pd.DataFrame):
+        if not isinstance(df, pd.DataFrame):
+            raise ValueError(f"data must be a pd.DataFrame, got {type(df)}")
+        if "item_id" not in df.columns:
+            raise ValueError(f"data must have a `item_id` column")
+        if "timestamp" not in df.columns:
+            raise ValueError(f"data must have a `timestamp` column")
+        if df["item_id"].isnull().any():
+            raise ValueError(f"`item_id` columns can not have nan")
+        if df["timestamp"].isnull().any():
+            raise ValueError(f"`timestamp` columns can not have nan")
+        if not df["item_id"].dtype == "int64":
+            raise ValueError(f"for item_id, the only pandas dtype allowed is ‘int64’.")
+        if not df["timestamp"].dtype == "datetime64[ns]":
+            raise ValueError(
+                f"for timestamp, the only pandas dtype allowed is ‘datetime64[ns]’."
+            )
+
+    @classmethod
+    def _validate_list_dataset(cls, list_dataset):
```

Comment:
Shouldn't we put this on `ListDataset`?

Then we can do `list_dataset.validate()`.

Reply:
Good point, will move it inside ListDataset.

Reply:
Actually I disagree with this point, going back to the design, and maybe will walk back some of my earlier discussions. Our original objective with the `ListDataset` object was that someone could bring their GluonTS data set and work out of the box with AG Forecasting. We seem to be headed in the other direction here by making `TimeSeriesListDataset` a concrete object,  also with checks like `isinstance(.., TimeSeriesListDataset`, and encapsulating its validation logic. This would make sense if we wanted this object to live and be passed around the library. We in fact want it to disappear.

We can revisit this in another PR.

Reply:
Hi Caner, it is clear to me that our core data structure will be `TimeSeriesDataFrame`. Having the validation logic inside `TimeSeriesListDataset` is very far from passing around in our library IMHO: 
1) I guess we will make it explicit that the forecasting predictor will require a `TimeSeriesDataFrame` to work. 
2) We have no plan to support data split or other functions in `TimeSeriesListDataset`.
3) When `TimeSeriesListDataset` disappears, it is actually easier to delete it as the validation is part of the `TimeSeriesListDataset` class.

Reply:
@canerturkmen @jaheba If we don't want to introduce a new class but to provide easy conversion for ListDataset in gluonts, one thing we could do is:
```
from collections.abc import Iterable

# in the __init__ of TimeSeriesDataFrame
def __init__(self, data: Any, *args, **kwargs):
    if isinstance(data, Iterable):
        data = self.from_list_dataset(data)    
    ...
```
Then we need to write in the documentation what format we are expecting and validate it in `self.from_list_dataset` method. One caveat is that we will assume the `start` timestamp has `freq`.

Reply:
Got me thinking, why do we want to subclass `DataFrame`? Couldn't we create a new class which uses a `DataFrame` internally instead?

Reply:
@huibinshen as @jaheba mentioned in the call, I think this problem is solved if we simply check for `DataFrame` first.

@jaheba this was discussed and decided in the design review. We want to carry over the data frame API; with all of its functions to enable ease of use for the customer in a non-obstrusive way and 2/adhere as much as possible to the rest of autogluon which uses simple pandas dataframes. So in this case, inheritance over composition.

I'm happy to take care of the validation + type issue in a future PR.

Reply:
thanks for the discussion. I'll resolve the conversation for now.